extensions/security/iprange/src/org/exist/security/realm/iprange/IPRangeRealm.java

```diff
@@ -0,0 +1,154 @@
+import java.util.Optional;
+import java.util.Properties;
+
+/**
+ * @author <a href="mailto:wshager@gmail.com">Wouter Hager</a>
+ */
+@ConfigurationClass("realm") //TODO: id = IPRange
+public class IPRangeRealm extends AbstractRealm {
+
+    @ConfigurationFieldAsAttribute("id")
+    public final static String ID = "IPRange";
+
+    @ConfigurationFieldAsAttribute("version")
+    public final static String version = "1.0";
+
+    protected final static Logger LOG = LogManager.getLogger(IPRangeRealm.class);
+    protected static IPRangeRealm instance = null;
+
+    public IPRangeRealm(final SecurityManagerImpl sm, final Configuration config) throws ConfigurationException {
+        super(sm, config);
+        instance = this;
+    }
+
+    private static long ipToLong(final InetAddress ip) {
+        final byte[] octets = ip.getAddress();
+        long result = 0;
+        for (final byte octet : octets) {
+            result <<= 8;
+            result |= octet & 0xff;
+        }
+        return result;
+    }
+
+	/*@Override
+    public void start(DBBroker broker) throws EXistException {
+		super.start(broker);
+	}*/
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    public boolean deleteAccount(final Account account) throws PermissionDeniedException, EXistException, ConfigurationException {
+        // Auto-generated method stub
+        return false;
+    }
+
+    @Override
+    public boolean deleteGroup(final Group group) throws PermissionDeniedException, EXistException, ConfigurationException {
+        // Auto-generated method stub
+        return false;
+    }
+
+    @Override
+    public Subject authenticate(final String ipAddress, final Object credentials) throws AuthenticationException {
+
+        // Elevaste to system privileges
+        try (final DBBroker broker = BrokerPool.getInstance().get(Optional.of(getSecurityManager().getSystemSubject()))) {
+
+            // Convert IP address
+            final long ipToTest = ipToLong(InetAddress.getByName(ipAddress));
+
+            // Get xquery service
+            final XQuery queryService = broker.getBrokerPool().getXQueryService();
+            if (queryService == null) {
+                LOG.error("IPRange broker unable to retrieve XQueryService");
+                return null;
+            }
+
+            // Construct XQuery
+            final String query = "collection('/db/system/security/IPRange/accounts')/account/" +
+                    "iprange[" + ipToTest + " ge number(start) and " + ipToTest + " le number(end)]/../name";
+            final XQueryContext context = new XQueryContext(broker.getBrokerPool());
+
+            final CompiledXQuery compiled = queryService.compile(broker, context, query);
+
+            final Properties outputProperties = new Properties();
+
+            // Execute xQuery
+            final Sequence result = queryService.execute(broker, compiled, null, outputProperties);
```

Comment:
This is maybe a comment for a future improvement:

At the moment you execute an XQuery upon every authenticate request. I should imagine that is not going to be very performant. It would be better to load the configuration from the `/db/system/security/IPRange/accounts` at startup or on modification.

A quick improvement would be to cache the compiled query, at the moment the query is compiled on every call.